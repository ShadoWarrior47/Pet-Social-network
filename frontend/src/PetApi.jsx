import axios from "axios";
// console.log("api.js loaded successfully");

// const BASE_URL = import.meta.env.VITE_API_BASE_URL || "http://localhost:3001";

const BASE_URL = "http://localhost:3001";

/** API Class.
 *
 * Static class tying together methods used to get/send to to the API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't
 * be any API-aware stuff elsewhere in the frontend.
 *
 */

class PetApi {
    // the token for interactive with the API will be stored here.
    static token = localStorage.getItem("token") || null;

    static setToken(token) {
        this.token = token;
        localStorage.setItem("token", token);
    }

    static getToken() {
        return localStorage.getItem("token");
    }

    static async request(endpoint, data = {}, method = "get") {
        console.debug("API Call:", endpoint, data, method);


        //there are multiple ways to pass an authorization token, this is how you pass it in the header.
        //this has been provided to show you another way to pass the token. you are only expected to read this code for this project.
        // Log the token to ensure it is being set
        // console.log("Token being sent:", this.token);

        const url = `${BASE_URL}/${endpoint}`;

        const headers = { Authorization: `Bearer ${this.token}` };
        const params = (method === "get")
            ? data
            : {};

        try {
            return (await axios({ url, method, data, params, headers })).data;
        } catch (err) {
            console.error("API Error:", err.response || err);

            let message = err.response?.data?.error?.message || "Unknown error occured";
            throw Array.isArray(message) ? message : [message];
        }
    }

    // Individual API routes

    /** Login user and get token. */
    static async login(data) {
        let res = await this.request("auth/token", data, "post");
        // this.setToken(res.token);
        return res.token;
    }

    /** Register a new user. */
    static async signup(data) {
        let res = await this.request("auth/register", data, "post");
        // this.setToken(res.token);
        return res.token;
    }

    /** Get user profile. */
    static async getCurrentUser(username) {
        let res = await this.request(`users/${username}`);
        return res.user;
    }

    /** Update user profile. */
    static async updateUser(username, data) {
        let res = await this.request(`users/${username}`, data, "patch");
        return res.user;
    }

    /** Delete an user by id. */
    static async deleteUser(username) {
        await this.request(`users/${username}`, {}, "delete");
    }

    /** Get GoogleAuthUrl. */
    static async getGoogleAuthUrl() {
        const res = await this.request("auth/google/url");
        return res.url;
    }

    /** Get all posts created by the logged-in user. */
    static async getUserPosts(username) {
        try {
            const res = await this.request(`users/${username}/posts`);
            return res.posts;
        } catch (err) {
            console.error("Error fetching user posts:", err);
            throw err;
        }
    }

    /** Get all pets created by the logged-in user. */
    static async getUserPets(username) {
        try {
            const res = await this.request(`users/${username}/pets`);
            return res.pets;
        } catch (err) {
            console.error("Error fetching user pets:", err);
            throw err;
        }
    }

    /** Create a new pet */
    static async createPet(petData) {
        try {
            const res = await this.request("pets", petData, "post");
            return res.pet;
        } catch (error) {
            console.error("Error in createPet:", error);
            throw error;
        }
    }

    /** Get a list of pets (with optional filters). */
    static async getPets(searchTerms = {}) {
        console.log("serachTarms before :",searchTerms);
        let res = await this.request("pets", searchTerms);
        return res.pets;
    }

    /** Get details on a pet by id. */
    static async getPet(id) {
        let res = await this.request(`pets/${id}`);
        return res.pet;
    }

    /** Get distinct pet types. */
    static async getDistinctTypes() {
        const res = await this.request("pets/types");
        return res.types;
    }

    /** Get distinct pet breeds. */
    static async getDistinctBreeds() {
        const res = await this.request("pets/breeds");
        return res.breeds;
    }

    /** Update an existing pet. */
    static async updatePet(id, petData) {
        const res = await this.request(`pets/${id}`, petData, "patch");
        // console.log(res.pet);
        return res.pet;
    }

    /** Delete a pet by ID. */
    static async deletePet(id) {
        await this.request(`pets/${id}`, {}, "delete");
    }

    /** Create a new post */
    static async createPost(postData) {
        try {
            const res = await this.request("posts", postData, "post");
            return res.post;
        } catch (error) {
            console.error("Error in createPost:", error);
            throw error;
        }
    }

    /** Get all posts (with optional filters). */
    static async getPosts(filters = {}) {
        const cleanedFilters = Object.fromEntries(
            Object.entries(filters).filter(([_, value]) => value && value.toString().trim() !== "")
        );

        try {
            const res = await this.request("posts", cleanedFilters);
            return res.posts;
        } catch (err) {
            console.error("Error fetching posts:", err);
            throw err;
        }
    }

    /** Get a specific post by id. */
    static async getPost(id) {
        let res = await this.request(`posts/${id}`);
        return res.post;
    }

    /** Update an existing post. */
    static async updatePost(id, postData) {
        const res = await this.request(`posts/${id}`, postData, "patch");
        // console.log(res.post);
        return res.post;
    }

    /** Delete a post by ID. */
    static async deletePost(id) {
        await this.request(`posts/${id}`, {}, "delete");
    }

    /** Create a new event. */
    static async createEvent(eventData) {
        const res = await this.request("events", eventData, "post");
        return res.event;
    }

    /** Get a list of events (with optional filters). */
    static async getEvents(filters = {}) {
        const cleanedFilters = Object.fromEntries(
            Object.entries(filters).filter(([_, value]) => {
                return typeof value === 'string' ? value.trim() !== "" : value != null;
            })
        );
        console.log("Clened Filters:", cleanedFilters);
        try {
            const res = await this.request("events", cleanedFilters);
            return res.events;
        } catch (err) {
            console.error("Error fetching events:", err);
            throw err;
        }
    }

    /** Get details on an event by id. */
    static async getEvent(id) {
        const res = await this.request(`events/${id}`);
        return res.event;
    }

    /** Update an existing event. */
    static async updateEvent(id, eventData) {
        const res = await this.request(`events/${id}`, eventData, "patch");
        return res.event;
    }

    /** Delete an event by id. */
    static async deleteEvent(id) {
        await this.request(`events/${id}`, {}, "delete");
    }


    /** Check is post liked by user.  */
    static async isPostLiked(postId) {
        try {
            const res = await this.request(`likes/${postId}`, {}, "get");
            return res.isLiked;
        } catch (err) {
            console.error("Error checking like status:", err);
            throw err;
        }
    }

    /** Get likes count. */
    static async getLikesCount(postId) {
        try {
            const res = await this.request(`likes/count/${postId}`, {}, "get");
            return res.likesCount;
        } catch (err) {
            console.error("Error fetching likes count:", err);
            throw err;
        }
    }


    /** Add like to post. */
    static async addLike(postId, data) {
        console.log("Adding like for post ID:", postId); // Debug log
        try {
            const res = await this.request("likes", { postId, ...data }, "post");
            console.log("Add Like Response:", res);
            return res;
        } catch (err) {
            console.error("Error adding like:", err.response || err);
            throw err;
        }
    }

    /** Remove like from post. */
    static async removeLike(postId) {
        try {
            return await this.request(`likes/${postId}`, {}, "delete");
        } catch (err) {
            console.error("Error removing like:", err);
            throw err;
        }
    }


    /** Get comments for post. */
    static async getComments(postId) {
        const res = await this.request(`comments/${postId}`, {}, "get");
        return res.comments;
    }

    /** Add comment to post. */
    static async addComment(postId, commentData) {
        const res = await this.request(`comments/${postId}`, commentData, "post");
        return res.comment;
    }


    /** Update an existing comment. */
    static async updateComment(postId, commentData) {
        const res = await this.request(`comments/${postId}`, commentData, "patch");
        console.log(res.comment);
        return res.comment;
    }

    /** Remove comment from post. */
    static async removeComment(postId) {
        console.log("Removing comment for post ID:", postId); // Debug log
        try {
            return await this.request(`comments/${postId}`, {}, "delete");
        } catch (err) {
            console.error("Error removing comment:", err);
            throw err;
        }
    }

    /** Get Notifications for user. */
    static async getNotifications(userId) {
        try {
            return await this.request(`notifications/${userId}`);
        } catch (error) {
            console.error("Error fetching notifications:", error);
            throw error;
        }
    }

    /** Mark Notifications as read. */
    static async markNotificationAsRead(notificationId) {
        try {
            return await this.request(`notifications/read/${notificationId}`, {}, 'post');
        } catch (error) {
            console.error("Error marking notification as read:", error);
            throw error;
        }
    }


}



export default PetApi;