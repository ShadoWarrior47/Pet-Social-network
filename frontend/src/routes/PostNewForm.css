/* General container styling */
.PostNewForm-container div {
    max-width: 600px !important;
    margin: 0px auto;
    padding: 10px;
    font-family: Arial, sans-serif;
    min-width: 500px;
}

.PostNewForm-container form {
    display: flex;
    flex-direction: column;
    gap: 10px;
    padding: 10px;
    background-color: #f9f9f9;
    border: 1px solid #ddd;
    border-radius: 5px;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.1);
}

.PostNewForm-container h1 {
    text-align: center;
    margin-bottom: 20px;
    font-size: 24px;
    color: #333;
}

.PostNewForm-container input {
    width: 100%;
    padding: 5px;
    font-size: 16px;
    border: 1px solid #ccc;
    border-radius: 4px;
    box-shadow: inset 0 1px 3px rgba(0, 0, 0, 0.1);
    outline: none;
    transition: border-color 0.3s;
}

.PostNewForm-container input:focus {
    border-color: #007bff;
}

.PostNewForm-container label {
    font-size: 14px;
    font-weight: bold;
    margin-bottom: 2px;
    color: #555;
    text-align: center;
}

.PostNewForm-container button {
    padding: 10px 20px;
    font-size: 16px;
    color: white;
    background-color: #007bff;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    transition: background-color 0.3s, transform 0.2s;
    width: 150px;
    margin: 0px auto;
}

.PostNewForm-container button:hover {
    background-color: #0056b3;
}

.PostNewForm-container button:active {
    transform: scale(0.98);
}