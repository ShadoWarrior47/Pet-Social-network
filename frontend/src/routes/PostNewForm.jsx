import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import PetApi from "../PetApi";
import "./PostNewForm.css";

function PostNewForm({ currentUser }) {
    const navigate = useNavigate();

    const INITIAL_STATE = {
        content: "",
        imageUrl: "",
    };

    const [formData, setFormData] = useState(INITIAL_STATE);

    const handleChange = (e) => {
        const { name, value } = e.target;
        setFormData((formData) => ({
            ...formData,
            [name]: value,
        }));
    };

    const handleSubmit = async (e) => {
        e.preventDefault();
        console.log("Form submitted");
        try {
            const postData = { ...formData, ownerId: currentUser.id };
            const createdPost = await PetApi.createPost(postData);

            if (createdPost && createdPost.id) {
                navigate(`/posts/${createdPost.id}`);
            } else {
                throw new Error("Post ID is missing in the response.");
            }
        } catch (err) {
            console.error("Error creating post:", err);
            alert("Failed to create post. Please try again.");
        }
    };

    return (
        <div className="PostNewForm-container">
            <h1>Add a New Post</h1>
            <form onSubmit={handleSubmit} role="form">
                <div>
                    <label htmlFor="content">Content :</label>
                    <input
                        id="content"
                        name="content"
                        type="text"
                        placeholder="Content"
                        value={formData.content}
                        onChange={handleChange}
                        required
                    />
                </div>
                <div>
                    <label htmlFor="imageUrl">Image URL :</label>
                    <input
                        id="imageUrl"
                        name="imageUrl"
                        type="text"
                        placeholder="Image URL"
                        value={formData.imageUrl}
                        onChange={handleChange}
                    />
                </div>
                <button type="submit">Add Post</button>
            </form>
        </div>
    );
}

export default PostNewForm;
