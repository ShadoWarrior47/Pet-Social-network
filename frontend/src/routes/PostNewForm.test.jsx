
import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import PostNewForm from "./PostNewForm";
import { vi } from "vitest";
import PetApi from "../PetApi";

// Mocking the PetApi module
vi.mock("../PetApi", () => ({
    default: {
      createPost: vi.fn(() => Promise.resolve({ id: 1 }))
    }
  }));
  
// Mocking the useNavigate hook
vi.mock('react-router-dom', async () => {
    const actual = await vi.importActual('react-router-dom');
    return {
      ...actual,
      useNavigate: () => vi.fn().mockImplementation(() => (path) => console.log(`Navigate to ${path}`)),
    };
  });

it("renders without crashing", function () {
    render(
        <MemoryRouter>
            <PostNewForm currentUser={{ id: 1 }} />
        </MemoryRouter>
    );
});

it("matches snapshot", function () {
    const { asFragment } = render(
        <MemoryRouter>
            <PostNewForm currentUser={{ id: 1 }} />
        </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

it("submits the form with post data", async function () {
    const { getByLabelText, getByText } = render(
        <MemoryRouter>
            <PostNewForm currentUser={{ id: 1 }} />
        </MemoryRouter>
    );

    const contentInput = getByLabelText("Content :");
    const imageUrlInput = getByLabelText("Image URL :");
    const submitButton = getByText("Add Post");

    // Simulate user typing into the input fields
    fireEvent.change(contentInput, { target: { value: "New Post Content" } });
    fireEvent.change(imageUrlInput, { target: { value: "http://example.com/image.jpg" } });

    // Simulate form submission
    fireEvent.click(submitButton);

    // Wait for expectations to avoid act warnings
    await waitFor(() => {
      expect(PetApi.createPost).toHaveBeenCalledWith({
        content: "New Post Content",
        imageUrl: "http://example.com/image.jpg",
        ownerId: 1
      });
    });
});



// import React from "react";
// import { render, fireEvent } from "@testing-library/react";
// import { MemoryRouter } from "react-router-dom";
// import PostNewForm from "./PostNewForm";
// import { vi } from "vitest";

// const mockCreatePost = vi.fn();

// // Mock react-router-dom
// const mockNavigate = vi.fn();
// vi.mock("react-router-dom", () => ({
//     ...vi.importActual("react-router-dom"),
//     useNavigate: () => mockNavigate,
// }));

// // Mock PetApi module
// vi.mock("../PetApi", () => ({
//     createPost: mockCreatePost,
// }));


// const currentUser = { id: 1 };

// describe("PostNewForm Tests", () => {
//     it("renders without crashing", function () {
//         render(
//             <MemoryRouter>
//                 <PostNewForm currentUser={currentUser} />
//             </MemoryRouter>
//         );
//     });

//     it("matches snapshot", function () {
//         const { asFragment } = render(
//             <MemoryRouter>
//                 <PostNewForm currentUser={currentUser} />
//             </MemoryRouter>
//         );
//         expect(asFragment()).toMatchSnapshot();
//     });

//     it("updates input fields correctly", function () {
//         const { getByLabelText } = render(
//             <MemoryRouter>
//                 <PostNewForm currentUser={currentUser} />
//             </MemoryRouter>
//         );

//         const contentInput = getByLabelText("Content :");
//         const imageUrlInput = getByLabelText("Image URL :");

//         fireEvent.change(contentInput, { target: { value: "New post content" } });
//         fireEvent.change(imageUrlInput, { target: { value: "http://example.com/image.jpg" } });

//         expect(contentInput.value).toBe("New post content");
//         expect(imageUrlInput.value).toBe("http://example.com/image.jpg");
//     });

//     it("submits form with correct data", async function () {
//         // mockCreatePost.mockResolvedValue({ id: 123 });

//         const { getByLabelText, getByText } = render(
//             <MemoryRouter>
//                 <PostNewForm currentUser={currentUser} />
//             </MemoryRouter>
//         );

//         fireEvent.change(getByLabelText("Content :"), { target: { value: "Test post" } });
//         fireEvent.change(getByLabelText("Image URL :"), { target: { value: "http://example.com/image.jpg" } });

//         fireEvent.click(getByText("Add Post"));

//         // Obtain the mock from PetApi
//         const { createPost } = require("../PetApi");
//         expect(createPost).toHaveBeenCalledWith({
//             content: "Test post",
//             imageUrl: "http://example.com/image.jpg",
//             ownerId: 1,
//         });

//         // expect(mockCreatePost).toHaveBeenCalledWith({
//         //     content: "Test post",
//         //     imageUrl: "http://example.com/image.jpg",
//         //     ownerId: 1,
//         // });

//         await expect(mockNavigate).toHaveBeenCalledWith("/posts/123");
//     });

//     it("shows alert on post creation failure", async function () {
//         // Update the mock to simulate a failure
//         const { createPost } = require("../PetApi");
//         createPost.mockRejectedValue(new Error("Failed to create post"));
//         window.alert = vi.fn();

//         const { getByText } = render(
//             <MemoryRouter>
//                 <PostNewForm currentUser={currentUser} />
//             </MemoryRouter>
//         );

//         fireEvent.click(getByText("Add Post"));

//         await expect(window.alert).toHaveBeenCalledWith("Failed to create post. Please try again.");
//     });
// });
