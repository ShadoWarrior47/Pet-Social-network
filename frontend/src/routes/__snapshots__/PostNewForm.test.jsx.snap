// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="PostNewForm-container"
  >
    <h1>
      Add a New Post
    </h1>
    <form
      role="form"
    >
      <div>
        <label
          for="content"
        >
          Content :
        </label>
        <input
          id="content"
          name="content"
          placeholder="Content"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="imageUrl"
        >
          Image URL :
        </label>
        <input
          id="imageUrl"
          name="imageUrl"
          placeholder="Image URL"
          type="text"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Add Post
      </button>
    </form>
  </div>
</DocumentFragment>
`;
